### PR TITLE
`mod intra_edge`: Backport memory reduction from dav1d 1.2.1

### PR DESCRIPTION
**Pull request by fbossen**
Eliminate duplicate and constant `intra_edge` table entries

### CODE REVIEW

**Comment by fbossen** [comment]
Relates to #286

File: src/intra_edge.rs [hunk 1]
```diff
@@ -18,18 +18,23 @@ bitflags! {
...
+    pub const TOP_HAS_RIGHT: Self = Self::union_all([
         Self::I444_TOP_HAS_RIGHT,
         Self::I422_TOP_HAS_RIGHT,
         Self::I420_TOP_HAS_RIGHT,
     ]);
 
+    pub const EDGE_ALL_TR_AND_BL: Self =
+        Self::union_all([Self::LEFT_HAS_BOTTOM, Self::TOP_HAS_RIGHT]);
+
+    pub const EDGE_NONE: Self = Self::empty();
```
Comment (folkertdev) [inline src/intra_edge.rs:36]: I think we can drop the `EDGE_` prefix from these constants. They'll always occur qualified as `EdgeFlags::*` so the edge bit is just repeated.

File: src/intra_edge.rs [hunk 2]
```diff
@@ -164,62 +164,27 @@ impl EdgeBranch {
...
-                    .select(matches!(bl, BlockLevel::Bl16x16)),
-            ),
-            EdgeFlags::TOP_HAS_RIGHT,
-            edge_flags.intersection(EdgeFlags::TOP_HAS_RIGHT),
-        ];
+        let h4 = EdgeFlags::LEFT_HAS_BOTTOM.union(
+            edge_flags
+                .intersection(EdgeFlags::I420_TOP_HAS_RIGHT)
+                .select(matches!(bl, BlockLevel::Bl16x16)),
+        );
```
Comment (folkertdev) [inline src/intra_edge.rs:171]: the convention we went with here is that variables can use `union`, so `foo.union(...)`, but when the left-hand side is an enum variant we use `union_all`:

```suggestion
        let h4 = EdgeFlags::union_all([
            EdgeFlags::LEFT_HAS_BOTTOM,
            edge_flags
                .intersection(EdgeFlags::I420_TOP_HAS_RIGHT)
                .select(matches!(bl, BlockLevel::Bl16x16)),
        ]);
```

File: src/intra_edge.rs [hunk 3]
```diff
@@ -164,62 +164,27 @@ impl EdgeBranch {
...
-            EdgeFlags::empty(),
-        ];
+        let v4 = EdgeFlags::TOP_HAS_RIGHT.union(
+            edge_flags
+                .intersection(EdgeFlags::union_all([
+                    EdgeFlags::I420_LEFT_HAS_BOTTOM,
+                    EdgeFlags::I422_LEFT_HAS_BOTTOM,
+                ]))
+                .select(matches!(bl, BlockLevel::Bl16x16)),
+        );
```
Comment (folkertdev) [inline src/intra_edge.rs:180]: like above

```suggestion
        let v4 = EdgeFlags::union_all([
            EdgeFlags::TOP_HAS_RIGHT,
            edge_flags
                .intersection(EdgeFlags::union_all([
                    EdgeFlags::I420_LEFT_HAS_BOTTOM,
                    EdgeFlags::I422_LEFT_HAS_BOTTOM,
                ]))
                .select(matches!(bl, BlockLevel::Bl16x16)),
        ]);
```

**Review by folkertdev — COMMENTED** [review]
I think this looks good except for some minor style/consistency nitpicks

**Comment by fbossen** [comment]
> I think this looks good except for some minor style/consistency nitpicks

Thanks for the feedback. I made the suggested changes.

**Review by folkertdev — APPROVED** [review]

File: src/intra_edge.rs [hunk 4]
```diff
@@ -164,62 +163,29 @@ impl EdgeBranch {
...
-            edge_flags.union(EdgeFlags::TOP_HAS_RIGHT),
-            EdgeFlags::empty(),
-        ];
+            edge_flags
+                .intersection(EdgeFlags::union_all([
+                    EdgeFlags::I420_LEFT_HAS_BOTTOM,
+                    EdgeFlags::I422_LEFT_HAS_BOTTOM,
+                ]))
+                .select(matches!(bl, BlockLevel::Bl16x16)),
+        ]);
```
Comment (kkysen) [inline src/intra_edge.rs:181]:
```suggestion
        let v4 = edge_flags
            .intersection(EdgeFlags::union_all([
                EdgeFlags::I420_LEFT_HAS_BOTTOM,
                EdgeFlags::I422_LEFT_HAS_BOTTOM,
            ]))
            .select(matches!(bl, BlockLevel::Bl16x16))
            .union(EdgeFlags::TOP_HAS_RIGHT),
```

File: src/intra_edge.rs [hunk 5]
```diff
@@ -18,18 +18,22 @@ bitflags! {
 }
 
 impl EdgeFlags {
-    const LEFT_HAS_BOTTOM: Self = Self::union_all([
+    pub const LEFT_HAS_BOTTOM: Self = Self::union_all([
```
Comment (kkysen) [inline src/intra_edge.rs:21]:
```suggestion
    pub const ALL_LEFT_HAS_BOTTOM: Self = Self::union_all([
```

I should've done this when I added this, but could add this in a separate commit on this PR?

File: src/intra_edge.rs [hunk 6]
```diff
@@ -18,18 +18,22 @@ bitflags! {
...
         Self::I444_LEFT_HAS_BOTTOM,
         Self::I422_LEFT_HAS_BOTTOM,
         Self::I420_LEFT_HAS_BOTTOM,
     ]);
 
-    const TOP_HAS_RIGHT: Self = Self::union_all([
+    pub const TOP_HAS_RIGHT: Self = Self::union_all([
```
Comment (kkysen) [inline src/intra_edge.rs:27]:
```suggestion
    pub const ALL_TOP_HAS_RIGHT: Self = Self::union_all([
```

Same reasoning and commit as above.

**Review by kkysen — CHANGES_REQUESTED** [review]
Mostly looks very good, but I had a few more suggestions.  Also, by using 2-byte `EdgeIndex`s instead of 8-byte pointers, our table should be even smaller than C's 🎉 .

**Review by kkysen — APPROVED** [review]

**Pull request merged**